### PR TITLE
Add flatpak mask command

### PR DESCRIPTION
**Pull request by alexlarsson**
This commands lets you selectively disable auto-download of extensions based on patterns. With this we can have extensions that install by default, yet still allow the user to not have them re-installed each time flatpak update is run.
    
This fixes https://github.com/flatpak/flatpak/issues/3090

### CODE REVIEW

File: common/flatpak-dir.c [hunk 1]
```diff
@@ -13558,6 +13591,56 @@ add_related (FlatpakDir *self,
...
+                    {
+                      if (i != 0)
+                        g_string_append (deny_regexp, "|");
+                      g_string_append (deny_regexp, regexp);
+                    }
+                }
+            }
+
+          g_string_append (deny_regexp, ")$");
+          g_print ("Masked regex: %s\n", deny_regexp->str);
```
Comment (matthiasclasen) [inline common/flatpak-dir.c]: Leftover debug spew ?

**Comment by matthiasclasen** [comment]
The new command needs a man page

**Comment by mwleeds** [comment]
I wonder if we could use this command to also fix https://github.com/flatpak/flatpak/issues/3078 which seems like a similar use case.

**Comment by Erick555** [comment]
As I understand `mask` will only block updates, so if someone has app/extension/whatever installed it will work like pinning current version.

**Comment by mwleeds** [comment]
Looks like as currently written it only applies to extensions and makes them invisible when an update operation is being calculated. So we'd have to add support for apps and runtimes.

**Comment by alexlarsson** [comment]
This doesn't really change anything for apps or runtimes, its only really about changing the default for the auto-download dependent extensions. For example, if you masked the openh264 extension but then manually installed it it would still update that.

The question then becomes, do we want this to *also* affect regular updates? Like, would you want to use this to keep the currently installed version of an app/runtime even across a "flatpak update" transaction. What about a more specific one, like "flatpak update org.the.masked.app"?

**Comment by alexlarsson** [comment]
New update, now with general masking.

File: doc/flatpak-mask.xml [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+        <variablelist>
+            <varlistentry>
+                <term><option>-h</option></term>
+                <term><option>--help</option></term>
+
+                <listitem><para>
+                    Show help options and exit.
+                </para></listitem>
+            </varlistentry>
+
```
Comment (matthiasclasen) [inline doc/flatpak-mask.xml]: You miss --version here, I think have that everywhere else.

File: doc/flatpak-mask.xml [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+    </refsynopsisdiv>
+
+    <refsect1>
+        <title>Description</title>
+
+        <para>
+          Flatpak maintains a list of patterns that that define which refs are masked.
+          A masked ref will never be updated or automatically installed (for example an
+          masked extension marked auto-download will not be downloaded). You can still
+          manually install such refs, but once they are installed the version will be pinned.
```
Comment (matthiasclasen) [inline doc/flatpak-mask.xml]: Just to make sure: you can manually install masked refs, but a manual update will not work ?

**Comment by matthiasclasen** [comment]
Thanks for adding the man page. One thing that is still missing is to add the new command to the list of commands in flatpak.1

File: doc/flatpak-mask.xml (hunk 3 above)
Comment (alexlarsson) [reply doc/flatpak-mask.xml]: Yeah, you can install it, but once its installed the current version is locked down. One could imagine being able to force-update with flatpak update if you specify the exact ref, but thats not going to work in general with libflatpak APIs, where you always specify a ref to update, so best be consistent here.

File: doc/flatpak-mask.xml (hunk 2 above)
Comment (alexlarsson) [reply doc/flatpak-mask.xml]: Really? I only see it in flatpak.xml

Comment (matthiasclasen) [reply doc/flatpak-mask.xml]: Sorry, i meant --verbose

Comment (alexlarsson) [reply doc/flatpak-mask.xml]: But --verbose is there.

**Comment by matthiasclasen** [comment]
Looks like some introspection error ?

**Pull request merged**